io_scene_godot/converters/armature.py

```diff
@@ -0,0 +1,96 @@
+"""Export a armature node"""
+import mathutils
+from ..structures import NodeTemplate
+
+
+def get_armature_data(node):
+    """Get the armature modifier of a blender object
+    if does not have one, return None"""
+    for modifier in node.modifiers:
+        if modifier.type.lower() == 'armature':
+            return modifier.object.data
+    return None
+
+
+class Bone:
+    """A Bone has almost same attributes as Godot bones"""
+
+    # must be ordered, Godot scene exporter force the first
+    # attribute of bone be 'name'
+    attributes = (
+        "name",
+        "parent",
+        "rest",
+        "pose",
+        "enabled",
+        "bound_children",
+    )
+
+    def __init__(self, bone_id, name, parent_id):
+        # name needs wrapped by double quotes
+        self.id = bone_id
+        self.name = '"{}"'.format(name)
+
+        # attributes
+        self.parent = parent_id
+        self.rest = mathutils.Matrix()
+        self.pose = mathutils.Matrix()
+        self.enabled = True
+        self.bound_children = None
+
+    def attr_to_key(self, attr_name):
+        """Add bone id to bone attribute"""
+        assert attr_name in Bone.attributes
+
+        return "bones/{}/{}".format(self.id, attr_name).lower()
+
+
+def export_bone_tree(pose_bone, parent_bone_id, bone_list):
```

Comment:
- the recursive is to track the parent_id(which is the id in escn file, not in the Blender bones structure), so if we want to rewrite it to iterative, we have to keep a map between Blender bone name and escn bone id, then when we export a bone we can set its parent id by look for bone.parent.name in the map.  However the map takes O(NlogN) time complexity, the current recursive implementation takes O(N) time complexity. So my opinion is, as long as the recursive depth is not exceeded, we use the recursive version.
- I think about the animation exporting, in my plan the pose of bone would be access through Blender data structure not escn tree, so it will not be a problem.